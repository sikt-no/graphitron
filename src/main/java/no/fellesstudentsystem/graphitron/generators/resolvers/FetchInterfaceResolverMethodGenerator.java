package no.fellesstudentsystem.graphitron.generators.resolvers;

import com.squareup.javapoet.CodeBlock;
import com.squareup.javapoet.MethodSpec;
import com.squareup.javapoet.TypeName;
import no.fellesstudentsystem.graphitron.definitions.fields.ObjectField;
import no.fellesstudentsystem.graphitron.definitions.objects.AbstractObjectDefinition;
import no.fellesstudentsystem.graphitron.definitions.objects.InterfaceDefinition;
import no.fellesstudentsystem.graphitron.definitions.objects.ObjectDefinition;
import no.fellesstudentsystem.graphitron.generators.abstractions.ResolverMethodGenerator;
import no.fellesstudentsystem.graphitron.generators.dependencies.Dependency;
import no.fellesstudentsystem.graphitron.generators.dependencies.QueryDependency;
import no.fellesstudentsystem.graphitron.schema.ProcessedSchema;
import no.fellesstudentsystem.graphitron.generators.abstractions.DBClassGenerator;

import java.util.Comparator;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

import static no.fellesstudentsystem.graphitron.generators.db.FetchDBClassGenerator.SAVE_DIRECTORY_NAME;
import static no.fellesstudentsystem.graphitron.mappings.JavaPoetClassName.*;
import static org.apache.commons.lang3.StringUtils.capitalize;
import static org.apache.commons.lang3.StringUtils.uncapitalize;

/**
 * Generates resolvers for queries returning an interface. E.g. the node resolver.
 */
public class FetchInterfaceResolverMethodGenerator extends ResolverMethodGenerator<ObjectField> {
    private final ObjectDefinition localObject;

    public FetchInterfaceResolverMethodGenerator(ObjectDefinition localObject, ProcessedSchema processedSchema) {
        super(localObject, processedSchema);
        this.localObject = localObject;
    }

    @Override
    public MethodSpec generate(ObjectField target) {
        InterfaceDefinition interfaceDefinition = processedSchema.getInterface(target);
        TypeName returnClassName = interfaceDefinition.getGraphClassName();

        var spec = getDefaultSpecBuilder(target.getName(), returnClassName);

        if (!localObject.isRoot()) {
            spec.addParameter(localObject.getGraphClassName(), uncapitalize(localObject.getName()));
        }

        var inputField = target.getInputFields().get(0);
        String inputFieldName = inputField.getName();
        spec
                .addParameter(inputIterableWrap(inputField), inputFieldName)
                .addParameter(DATA_FETCHING_ENVIRONMENT.className, ENV_NAME)
                .addStatement("String tablePartOfId = $T.getTablePartOf($N)", FIELD_HELPERS.className, inputFieldName)
                .addCode("\n");

        processedSchema
                .getObjects()
                .values()
                .stream()
                .filter(it -> it.implementsInterface(interfaceDefinition.getName()))
                .sorted(Comparator.comparing(AbstractObjectDefinition::getName))
                .forEach(implementation -> {
                    var queryLocation = implementation.getName() + DBClassGenerator.FILE_NAME_SUFFIX;
                    var map = Map.ofEntries(
                            Map.entry("env", ENV_NAME),
                            Map.entry("string", STRING.className),
                            Map.entry("returnType", returnClassName),
                            Map.entry("dataloader", DATA_LOADER_FACTORY.className),
                            Map.entry("batchLoader", MAPPED_BATCH_LOADER_WITH_CONTEXT.className),
                            Map.entry("implementationClass", implementation.getGraphClassName()),
                            Map.entry("future", COMPLETABLE_FUTURE.className),
                            Map.entry("queryInstanceField", uncapitalize(queryLocation)),
                            Map.entry("implName", implementation.getName()),
                            Map.entry("inputFieldName", inputFieldName),
                            Map.entry("inputFieldNameCap", capitalize(inputFieldName)),
                            Map.entry("selectionSets", SELECTION_SET.className),
                            Map.entry("environmentUtils", ENVIRONMENT_UTILS.className),
                            Map.entry("referenceFieldName", capitalize(target.getName())),
                            Map.entry("context", Dependency.CONTEXT_NAME)
                    );
                    dependencySet.add(new QueryDependency(queryLocation, SAVE_DIRECTORY_NAME));

                    spec.beginControlFlow("if (tablePartOfId.equals($T.$N.getViewId().toString()))", TABLES.className, implementation.getTable().getMappingName());
                    spec.addStatement(CodeBlock.builder().addNamed(
                                    "return $env:N.getDataLoaderRegistry().<$string:T, $returnType:T>computeIfAbsent(tablePartOfId, name ->$W" +
                                            "$dataloader:T.newMappedDataLoader(($batchLoader:T<$string:T, $implementationClass:T>) (keys, loaderEnvironment) ->$>$W" +
                                            "$future:T.completedFuture($queryInstanceField:N.load$implName:NBy$inputFieldNameCap:NsAs$referenceFieldName:N" +
                                            "($context:N, keys, new $selectionSets:T($environmentUtils:T.getSelectionSetsFromEnvironment(loaderEnvironment))))))" +
                                            "$<$W.load($inputFieldName:N, $env:N)", map)
                            .build());
                    spec.endControlFlow();
                });

        spec.addStatement("throw new $T(\"could not find dataloader for $N with prefix \" + tablePartOfId)", ILLEGAL_ARGUMENT_EXCEPTION.className, inputFieldName);

        return spec.build();
    }

    @Override
    public List<MethodSpec> generateAll() {
        return localObject
                .getFields()
                .stream()
                .filter(ObjectField::isGenerated)
                .filter(it -> processedSchema.isInterface(it.getTypeName()))
                .map(this::generate)
                .collect(Collectors.toList());
    }

    @Override
    public boolean generatesAll() {
        return localObject
                .getFields()
                .stream()
                .filter(it -> processedSchema.isInterface(it.getTypeName()))
                .allMatch(ObjectField::isGenerated);
    }
}
