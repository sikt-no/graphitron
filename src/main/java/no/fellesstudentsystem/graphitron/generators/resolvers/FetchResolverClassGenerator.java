package no.fellesstudentsystem.graphitron.generators.resolvers;

import com.squareup.javapoet.TypeSpec;
import no.fellesstudentsystem.graphitron.definitions.objects.ObjectDefinition;
import no.fellesstudentsystem.graphitron.generators.abstractions.ResolverClassGenerator;
import no.fellesstudentsystem.graphitron.schema.ProcessedSchema;

import java.io.IOException;
import java.util.*;
import java.util.stream.Collectors;

import static no.fellesstudentsystem.graphql.mapping.GraphQLReservedName.SCHEMA_ROOT_NODE_MUTATION;

/**
 * Class generator for basic select resolver classes.
 */
public class FetchResolverClassGenerator extends ResolverClassGenerator<ObjectDefinition> {
    public final String SAVE_DIRECTORY_NAME = ResolverClassGenerator.DEFAULT_SAVE_DIRECTORY_NAME + ".query";

    public FetchResolverClassGenerator(ProcessedSchema processedSchema) {
        super(processedSchema);
    }

    @Override
    public void generateQualifyingObjectsToDirectory(String path, String packagePath) throws IOException {
        var classes = processedSchema
                .getObjects()
                .values()
                .stream()
                .filter(ObjectDefinition::isGenerated)
                .filter(obj -> !obj.getName().equals(SCHEMA_ROOT_NODE_MUTATION.getName()))
                .map(this::generate)
                .collect(Collectors.toList());

        for (var generatedClass : classes) {
            writeToFile(generatedClass, path, packagePath, SAVE_DIRECTORY_NAME);
        }
    }

    @Override
    public TypeSpec generate(ObjectDefinition target) {
        return getSpec(
                target.getName(),
                List.of(
                        new FetchResolverMethodGenerator(target, processedSchema),
                        new FetchInterfaceResolverMethodGenerator(target, processedSchema)
                )
        ).build();
    }

    @Override
    public String getDefaultSaveDirectoryName() {
        return SAVE_DIRECTORY_NAME;
    }
}
