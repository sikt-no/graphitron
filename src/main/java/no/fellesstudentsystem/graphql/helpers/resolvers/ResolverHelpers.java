package no.fellesstudentsystem.graphql.helpers.resolvers;

import graphql.schema.DataFetchingEnvironment;
import no.fellesstudentsystem.graphql.helpers.EnvironmentUtils;
import no.fellesstudentsystem.graphql.helpers.selection.SelectionSet;
import org.jetbrains.annotations.NotNull;
import org.jooq.DSLContext;

import java.util.Optional;

public class ResolverHelpers {
    public static int getPageSize(Integer first, int max, int defaultMax) {
        return Optional.ofNullable(first).map(it -> Math.min(max, it)).orElse(defaultMax);
    }

    public static DSLContext selectContext(DataFetchingEnvironment env, DSLContext defaultContext) {
        return env.getLocalContext() == null ? defaultContext : (DSLContext) env.getLocalContext();
    }

    @NotNull
    public static SelectionSet getSelectionSet(DataFetchingEnvironment env) {
        return new SelectionSet(EnvironmentUtils.getSelectionSetsFromEnvironment(env));
    }
}
