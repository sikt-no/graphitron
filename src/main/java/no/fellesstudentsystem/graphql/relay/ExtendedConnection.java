package no.fellesstudentsystem.graphql.relay;

import graphql.relay.Connection;

import java.util.List;

public interface ExtendedConnection<T> extends Connection<T> {

    List<T> getNodes();

    Integer getTotalCount();
}
