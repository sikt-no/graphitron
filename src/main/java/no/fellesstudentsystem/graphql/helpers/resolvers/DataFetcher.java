package no.fellesstudentsystem.graphql.helpers.resolvers;

import graphql.ErrorType;
import graphql.GraphqlErrorBuilder;
import graphql.schema.DataFetchingEnvironment;
import no.fellesstudentsystem.graphql.exception.ValidationViolationGraphQLException;
import no.fellesstudentsystem.graphql.helpers.EnvironmentUtils;
import no.fellesstudentsystem.graphql.helpers.functions.DBCount;
import no.fellesstudentsystem.graphql.helpers.functions.DBQuery;
import no.fellesstudentsystem.graphql.helpers.functions.DBQueryRoot;
import no.fellesstudentsystem.graphql.helpers.selection.SelectionSet;
import no.fellesstudentsystem.graphql.relay.ConnectionImpl;
import org.apache.commons.lang3.tuple.Pair;
import org.dataloader.DataLoaderFactory;
import org.dataloader.MappedBatchLoaderWithContext;
import org.jooq.DSLContext;

import java.util.*;
import java.util.concurrent.CompletableFuture;
import java.util.function.Function;
import java.util.stream.Collectors;

import static no.fellesstudentsystem.graphql.naming.GraphQLReservedName.CONNECTION_TOTAL_COUNT;

public class DataFetcher extends AbstractFetcher {
    /**
     * Create a dataloader for a resolver.
     * @param env Environment for the resolver.
     * @param ctx Context from the resolver.
     */
    public DataFetcher(DataFetchingEnvironment env, DSLContext ctx) {
        super(env, ctx);
    }

    /**
     * Load the data for a root resolver.
     * @param dbFunction Function to call to retrieve the query data.
     * @return A resolver result.
     * @param <T> Type that the resolver fetches.
     */
    public <T> CompletableFuture<T> load(DBQueryRoot<T> dbFunction) {
        return CompletableFuture.completedFuture(dbFunction.callDBMethod(this.ctx, selection));
    }

    /**
     * Load the data for a root resolver. The result is paginated.
     * @param pageSize Size of the pages for pagination.
     * @param maxNodes Limit on how many elements may be fetched at once.
     * @param dbFunction Function to call to retrieve the query data.
     * @param countFunction Function to call to retrieve the total count of elements that could be potentially retrieved.
     * @param connectionFunction Function that converts the result of the query to a GraphQL connection structure.
     * @return A paginated resolver result.
     * @param <T> Type that the resolver fetches.
     */
    public <T, U> CompletableFuture<U> loadPaginated(
            int pageSize,
            int maxNodes,
            DBQueryRoot<List<Pair<String, T>>> dbFunction,
            DBCount<String> countFunction,
            Function<ConnectionImpl<T>, U> connectionFunction
    ) {
        return CompletableFuture.completedFuture(
                getPaginatedConnection(
                        dbFunction.callDBMethod(ctx, connectionSelection),
                        pageSize,
                        connectionSelection.contains(CONNECTION_TOTAL_COUNT.getName()) ? countFunction.callDBMethod(ctx, Set.of()) : -1,
                        maxNodes,
                        connectionFunction
                )
        );
    }

    /**
     * Load the data for a resolver.
     * @param resolveName Name of the resolver.
     * @param dbFunction Function to call to retrieve the query data.
     * @param id ID of the queried element.
     * @return A resolver result.
     * @param <T> Type that the resolver fetches.
     */
    public <T> CompletableFuture<T> load(String resolveName, String id, DBQuery<String, T> dbFunction) {
        return getLoader(resolveName, (keys, set) -> getMappedDataLoader(keys, set, dbFunction))
                .load(asKeyPath(id), env);
    }

    /**
     * Load the data for a resolver. The result is paginated.
     * @param pageSize Size of the pages for pagination.
     * @param maxNodes Limit on how many elements may be fetched at once.
     * @param dbFunction Function to call to retrieve the query data.
     * @param countFunction Function to call to retrieve the total count of elements that could be potentially retrieved.
     * @param connectionFunction Function that converts the result of the query to a GraphQL connection structure.
     * @return A paginated resolver result.
     * @param <T> Type that the resolver fetches.
     */
    public <T, U> CompletableFuture<U> loadPaginated(
            String resolveName,
            String id,
            int pageSize,
            int maxNodes,
            DBQuery<String, List<Pair<String, T>>> dbFunction,
            DBCount<String> countFunction,
            Function<ConnectionImpl<T>, U> connectionFunction
    ) {
        return getConnectionLoader(resolveName, (keys, set) -> getMappedDataLoader(keys, set, maxNodes, pageSize, dbFunction, countFunction, connectionFunction))
                .load(asKeyPath(id), env);
    }

    /**
     * Load the data for a resolver.
     * @param resolveName Name of the resolver.
     * @param dbFunction Function to call to retrieve the query data.
     * @param id ID of the queried element.
     * @return A resolver result.
     * @param <T> Type that the resolver fetches.
     */
    public <T> CompletableFuture<List<T>> loadNonNullable(String resolveName, String id, DBQuery<String, List<T>> dbFunction) {
        return getLoader(resolveName, (keys, set) -> getMappedDataLoader(keys, set, dbFunction))
                .load(asKeyPath(id), env)
                .thenApply(data -> Optional.ofNullable(data).orElse(List.of()));
    }

    /**
     * Load the data for a resolver in a lookup format.
     * @param keys The lookup keys that are necessary for match rows in the result with the original resolver input.
     * @param dbFunction Function to call to retrieve the query data.
     * @return A resolver result without pagination.
     * @param <K> The key type that this lookup uses.
     * @param <V> Type that the resolver fetches.
     */
    public <K, V> CompletableFuture<List<V>> loadLookup(List<List<K>> keys, DBQuery<String, V> dbFunction) {
        if (keys.isEmpty()) {
            return CompletableFuture.completedFuture(List.of());
        }

        var mergedKeys = mergeKeys(keys, env);

        var dbResult = dbFunction.callDBMethod(ctx, new HashSet<>(mergedKeys), selection);
        var orderedResult = mergedKeys.stream().map(dbResult::get).collect(Collectors.toList());
        return CompletableFuture.completedFuture(orderedResult);
    }

    public static <K> List<String> mergeKeys(List<List<K>> keys, DataFetchingEnvironment env) {
        if (keys.isEmpty()) {
            return List.of();
        }

        // This assumes all keys are correlated.
        var nKeys = keys.get(0).size();
        if (keys.stream().map(List::size).anyMatch(it -> it != nKeys)) {
            throw new ValidationViolationGraphQLException(
                    List.of(
                            GraphqlErrorBuilder
                                    .newError(env)
                                    .path(env.getExecutionStepInfo().getPath())
                                    .message("Keys sets have differing lengths. For this type of query, each key field is required to be an array of equal length.")
                                    .errorType(ErrorType.ValidationError)
                                    .build()
                    )
            );
        }

        var mergedKeys = new ArrayList<>(Collections.nCopies(nKeys, ""));
        for (var keyList : keys) {
            for (int i = 0; i < nKeys; i++) {
                var currentKey = mergedKeys.get(i);
                mergedKeys.set(i, (currentKey.isEmpty() ? "" : currentKey + ",") + keyList.get(i));
            }
        }

        return mergedKeys;
    }

    /**
     * Load the data for an interface resolver.
     * @param table The source from which the data should be loaded.
     * @param keyToLoad The specific key that should be loaded.
     * @param dbFunction Function to call to retrieve the query data.
     * @return A resolver result without pagination.
     * @param <U> Type that the resolver fetches.
     */
    public <T, U extends T> CompletableFuture<T> loadInterface(String table, String keyToLoad, DBQuery<String, U> dbFunction) {
        return env
                .getDataLoaderRegistry()
                .<String, T>computeIfAbsent(table, name ->
                        DataLoaderFactory.newMappedDataLoader((MappedBatchLoaderWithContext<String, U>) (ids, loaderEnvironment) ->
                                CompletableFuture.completedFuture(dbFunction.callDBMethod(ctx, ids, new SelectionSet(EnvironmentUtils.getSelectionSetsFromEnvironment(loaderEnvironment))))
                        )
                )
                .load(keyToLoad, env);
    }

    private <T, U> CompletableFuture<Map<String, U>> getMappedDataLoader(
            Set<String> keys,
            SelectionSet selectionSet,
            int maxNodes,
            int pageSize,
            DBQuery<String, List<Pair<String, T>>> dbFunction,
            DBCount<String> countFunction,
            Function<ConnectionImpl<T>, U> connectionFunction
    ) {
        if (keys.isEmpty()) {
            return CompletableFuture.completedFuture(Map.of());
        }

        var keyToId = getKeyToId(keys);
        var idSet = new HashSet<>(keyToId.values());
        return CompletableFuture.completedFuture(
                getPaginatedConnection(
                        resultAsMap(keyToId, dbFunction.callDBMethod(ctx, idSet, selectionSet)),
                        pageSize,
                        connectionSelection.contains(CONNECTION_TOTAL_COUNT.getName()) ? countFunction.callDBMethod(ctx, idSet) : -1,
                        maxNodes,
                        connectionFunction
                )
        );
    }


    private <T> CompletableFuture<Map<String, T>> getMappedDataLoader(Set<String> keys, SelectionSet selectionSet, DBQuery<String, T> dbFunction) {
        if (keys.isEmpty()) {
            return CompletableFuture.completedFuture(Map.of());
        }

        var keyToId = getKeyToId(keys);
        return CompletableFuture.completedFuture(resultAsMap(keyToId, dbFunction.callDBMethod(ctx, new HashSet<>(keyToId.values()), selectionSet)));
    }
}
