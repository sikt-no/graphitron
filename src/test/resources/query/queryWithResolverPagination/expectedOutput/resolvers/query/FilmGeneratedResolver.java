package fake.code.generated.resolvers.query;

import fake.code.generated.queries.query.FilmDBQueries;
import fake.graphql.example.api.FilmResolver;
import fake.graphql.example.model.Film;
import fake.graphql.example.model.Inventory;
import graphql.schema.DataFetchingEnvironment;
import java.lang.Exception;
import java.lang.Integer;
import java.lang.Override;
import java.lang.String;
import java.util.concurrent.CompletableFuture;
import javax.inject.Inject;

import no.fellesstudentsystem.graphql.helpers.resolvers.DataFetcher;
import no.fellesstudentsystem.graphql.helpers.resolvers.ResolverHelpers;
import no.fellesstudentsystem.graphql.relay.ExtendedConnection;
import org.jooq.DSLContext;

public class FilmGeneratedResolver implements FilmResolver {
    @Inject
    DSLContext ctx;

    @Inject
    private FilmDBQueries filmDBQueries;

    @Override
    public CompletableFuture<ExtendedConnection<Inventory>> inventory(Film film, Integer first,
            String after, DataFetchingEnvironment env) throws Exception {
        int pageSize = ResolverHelpers.getPageSize(first, 1000, 10);
        return new DataFetcher(env, this.ctx).load("inventoryForFilm", film.getId(), pageSize, 1000,
                (ctx, ids, selectionSet) -> filmDBQueries.inventoryForFilm(ctx, ids, pageSize, after, selectionSet),
                (ctx, ids, selectionSet) -> selectionSet.contains("totalCount") ? filmDBQueries.countInventoryForFilm(ctx, ids) : null,
                (it) -> it.getId());
    }
}
