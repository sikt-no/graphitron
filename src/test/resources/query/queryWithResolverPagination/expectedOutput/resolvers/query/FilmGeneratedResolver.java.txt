package fake.code.generated.resolvers.query;

import fake.code.generated.queries.query.FilmDBQueries;
import fake.graphql.example.package.api.FilmResolver;
import fake.graphql.example.package.model.Film;
import fake.graphql.example.package.model.Inventory;
import graphql.relay.DefaultConnectionCursor;
import graphql.relay.DefaultEdge;
import graphql.relay.DefaultPageInfo;
import graphql.schema.DataFetchingEnvironment;
import java.lang.Exception;
import java.lang.Integer;
import java.lang.Math;
import java.lang.Override;
import java.lang.String;
import java.util.AbstractMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Optional;
import java.util.concurrent.CompletableFuture;
import java.util.stream.Collectors;
import javax.inject.Inject;
import no.fellesstudentsystem.graphql.helpers.EnvironmentUtils;
import no.fellesstudentsystem.graphql.helpers.selection.ConnectionSelectionSet;
import no.fellesstudentsystem.graphql.relay.ConnectionImpl;
import no.fellesstudentsystem.graphql.relay.ExtendedConnection;
import org.dataloader.DataLoader;
import org.dataloader.DataLoaderFactory;
import org.dataloader.MappedBatchLoaderWithContext;
import org.jooq.DSLContext;

public class FilmGeneratedResolver implements FilmResolver {
    @Inject
    DSLContext ctx;

    @Inject
    private FilmDBQueries filmDBQueries;

    @Override
    public CompletableFuture<ExtendedConnection<Inventory>> inventory(Film film, Integer first,
            String after, DataFetchingEnvironment env) throws Exception {
        var ctx = env.getLocalContext() == null ? this.ctx : (DSLContext) env.getLocalContext();
        int pageSize = Optional.ofNullable(first).orElse(10);
        DataLoader<String, ExtendedConnection<Inventory>> loader = env.getDataLoaderRegistry().computeIfAbsent("inventoryForFilm", name -> {
            var batchLoader = (MappedBatchLoaderWithContext<String, ExtendedConnection<Inventory>>) (keys, batchEnvLoader) -> {
                var keyToId = keys.stream().collect(
                        Collectors.toMap(s -> s, s -> s.substring(s.lastIndexOf("||") + 2)));
                var idSet = new HashSet<>(keyToId.values());
                var selectionSet = new ConnectionSelectionSet(EnvironmentUtils.getSelectionSetsFromEnvironment(batchEnvLoader));
                var dbResult = filmDBQueries.inventoryForFilm(ctx, idSet, pageSize, after, selectionSet);
                var totalCount = selectionSet.contains("totalCount") ? filmDBQueries.countInventoryForFilm(ctx, idSet) : null;
                var mapResult = keyToId.entrySet().stream()
                        .filter(it -> dbResult.get(it.getValue()) != null)
                        .collect(Collectors.toMap(Map.Entry::getKey, it -> dbResult.get(it.getValue())));
                var pagedResult = mapResult.entrySet().stream().map(resultEntry -> {
                    var resultValue = resultEntry.getValue();
                    var size = Math.min(resultValue.size(), pageSize);
                    var items = resultValue.subList(0, size);
                    var firstItem = items.size() == 0 ? null : new DefaultConnectionCursor(items.get(0).getId());
                    var lastItem = items.size() == 0 ? null : new DefaultConnectionCursor(items.get(items.size() - 1).getId());
                    var pagedResultEntry = ConnectionImpl
                            .<Inventory>builder()
                            .setPageInfo(
                                    new DefaultPageInfo(firstItem, lastItem, false, resultValue.size() > pageSize)
                            )
                            .setNodes(items)
                            .setEdges(
                                    items
                                            .stream()
                                            .map(item -> new DefaultEdge<Inventory>(item, new DefaultConnectionCursor(item.getId())))
                                            .collect(Collectors.toList())
                            )
                            .setTotalCount(totalCount != null ? Math.min(1000, totalCount) : null)
                            .build();
                    return new AbstractMap.SimpleEntry<String, ExtendedConnection<Inventory>>(resultEntry.getKey(), pagedResultEntry);
                } ).collect(Collectors.toMap(r -> r.getKey(), r -> r.getValue()));
                return CompletableFuture.completedFuture(pagedResult);
            } ;
            return DataLoaderFactory.newMappedDataLoader(batchLoader);
        } );
        return loader.load(env.getExecutionStepInfo().getPath().toString() + "||" + film.getId(), env);
    }
}